scipy/linalg/src/_npymath.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Helpers for dealing with npy_cdouble/npy_cfloat types.
+ */
+#pragma once
+#include "numpy/npy_math.h"
+
+namespace _numpymath{
+
+/*
+ * std::numeric_limits and useful constants.
+ */
+
+template<typename T> struct numeric_limits {};
+
+template<>
+struct numeric_limits<float>{
+    static constexpr double one = 1.0f;
+    static constexpr float nan = std::numeric_limits<float>::quiet_NaN();
+};
+
+template<>
+struct numeric_limits<double>{
+    static constexpr double one = 1.0;
+    static constexpr double nan = std::numeric_limits<double>::quiet_NaN();
+};
+
+
+template<>
+struct numeric_limits<npy_cfloat>{
+    static constexpr npy_cfloat one = {1.0f, 0.0f};
+    static constexpr npy_cfloat nan = {std::numeric_limits<float>::quiet_NaN(),
+                                       std::numeric_limits<float>::quiet_NaN()};
+};
+
+template<>
+struct numeric_limits<npy_cdouble>{
+    static constexpr npy_cdouble one = {1.0, 0.0};
+    static constexpr npy_cdouble nan = {std::numeric_limits<double>::quiet_NaN(),
+                                        std::numeric_limits<double>::quiet_NaN()};
+};
+
+
+
+/* 
+ * Grab a real part of a possibly complex array.
+ * This is for the work queries.
+ * There must be a better way, I'm sure.
```

Comment:
There isn't, this is the templating bloat I was mentioning before. Even to get the real part this is the boilerplate code we have to inject.

Reply:
Minor comment that it's not templating, really. It's just overloading, so would be exactly the same if it all were pure C.